docs/RingBuffer.md

```diff
@@ -0,0 +1,312 @@
+# New ebpf Ring Buffer Map (proposal)
+
+## Overview
+
+The current ringbuffer uses a pure callback-based approach to reading the ringbuffer.
+Linux also supports memory-mapped polling consumers, which can't be directly supported in the current model.
+
+The new ring buffer will use [I/O Completion Ports](https://learn.microsoft.com/en-us/windows/win32/fileio/i-o-completion-ports),
+which is the windows API closest to the linux epoll mechanism, and memory map producer+consumer pages into the user process in the same way that linux does.
+
+The new API will support 2 access methods: callbacks and mmap/epoll style access.
+
+callback consumer:
+
+1. call `ring_buffer__new` to set up callback
+2. call `ring_buffer__consume` as needed to invoke the callback on any outsanding data that is ready
+
+mmap/epoll consumer:
+
+1. call `ring_buffer__new` to get a ringbuffer manager
+2. call `ebpf_ring_buffer_get_buffer` to get pointers to the mapped producer/consumer pages
+3. directly read records from the producer pages (and update consumer offset as we read)
+4. call `ring_buffer__epoll_fd` to get the iocp handle as a fd_t so we can wait for new data
+5. use `GetQueuedCompletionStatus`/`GetQueuedCompletionStatusEx` to wait for new data to be available
+    - We could provide a wrapper to make the common ebpf use cases simpler
+
+## API Changes
+
+### Changes to ebpf helper functions
+
+```c
+/**
+ * @brief Output record to ringbuf
+ *
+ * Note newly added flags (to specify wakeup options).
+ *
+ * Wakeup options:
+ * - auto (default): Notify if consumer has caught up.
+ * - BPF_RB_FORCE_WAKEUP - Always notify consumer.
+ * - BPF_RB_NO_WAKEUP - Never notify consumer.
+ *
+ */
+ebpf_result_t
+ebpf_ring_buffer_output(ebpf_ring_buffer_t* ring, uint8_t* data, size_t length, size_t flags)
```

Comment:
just to confirm: is this changing the helper function signature for `bpf_ringbuf_output()`?

Reply:
Actually no. I don't know what I was looking at before thinking we didn't have flags, but we do so there is no change to the signature (I'll edit the md).